{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4c427245",
   "metadata": {},
   "source": [
    "# Strelka to Sc-TUSV-ext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e17927",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "from io import StringIO\n",
    "\n",
    "def convert_chromosome(chrom):\n",
    "    if chrom == 'chrX':\n",
    "        return 23\n",
    "    else:\n",
    "        return int(chrom[3:])\n",
    "\n",
    "def read_vcf_without_headers(input_file_path):\n",
    "    filtered_lines = []\n",
    "    with gzip.open(input_file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('##'):\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "    filtered_content = ''.join(filtered_lines)\n",
    "    df_snv = pd.read_csv(StringIO(filtered_content), sep='\\t')\n",
    "    df_snv = df_snv[df_snv['#CHROM'] != 'chrY']\n",
    "    df_snv['chr'] = df_snv['#CHROM'].apply(convert_chromosome)\n",
    "    df_snv['data'] = df_snv[list(df_snv.columns)[-2]]\n",
    "    display(df_snv)\n",
    "    return df_snv\n",
    "\n",
    "def strelka_to_sctusv(df_snv, output_file_path):\n",
    "    \n",
    "    file_vcf = open(output_file_path,'w')\n",
    "    file_vcf.write('##fileformat=VCFv4.2\\n\\\n",
    "##filedate=20211011\\n\\\n",
    "##INFO=<ID=END,Number=1,Type=Integer,Description=\"End position of the variant described in this record\">\\n\\\n",
    "##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description=\"Imprecise structural variation\">\\n\\\n",
    "##INFO=<ID=MATEID,Number=.,Type=String,Description=\"ID of mate breakends\">\\n\\\n",
    "##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">\\n\\\n",
    "##FORMAT=<ID=GT,Number=1,Type=Integer,Description=\"Genotype\">\\n\\\n",
    "##FORMAT=<ID=CN,Number=2,Type=Integer,Description=\"Copy number genotype for imprecise events\">\\n\\\n",
    "##FORMAT=<ID=CNADJ,Number=.,Type=Integer,Description=\"Copy number of adjacency\">\\n\\\n",
    "##FORMAT=<ID=BDP,Number=1,Type=Integer,Description=\"Depth of split reads\">\\n\\\n",
    "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read depth\">\\n\\\n",
    "##ALT=<ID=DEL,Description=\"Deletion\">\\n\\\n",
    "##ALT=<ID=DUP,Description=\"Duplication\">\\n\\\n",
    "##ALT=<ID=INS,Description=\"Insertion of novel sequence\">\\n\\\n",
    "##ALT=<ID=CNV,Description=\"Copy number variable region\">\\n\\\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tTUMOR\tNORMAL\\n')\n",
    "\n",
    "    \n",
    "\n",
    "    # Writing the CNVs\n",
    "    #count=0\n",
    "    #for index, row in df_dip_cnvs.iterrows():\n",
    "\n",
    "    #    line = str(row['chr'])+'\\t'+str(row['start'])+'\\t'+'cnv'+str(count)+'\\t'+'.'+'\\t'+'<CNV>'+'\\t'+'.\\tPASS'+'\\t'\\\n",
    "    #                    +'END='+str(row['end'])+';IMPRECISE'+'\\t'+'GT:CN\\t'+str('0|1')+':'+str(row['cn_a'])+','+str(row['cn_b'])+'\\t'+'0|0:1,1\\n'\n",
    "    #    file_vcf.write(line)\n",
    "    #    count+=1\n",
    "\n",
    "    \n",
    "    # Writing the SNVs\n",
    "    count=0\n",
    "    for index, row in df_snv.iterrows():\n",
    "        if len(str(row['REF']))==1 and len(str(row['ALT']))==1:        \n",
    "            line = str(row['chr'])+'\\t'+str(row['POS'])+'\\t'+'snv'+str(count)+'\\t'+row['REF']+'\\t'+row['ALT']+'\\t'+'.\\tPASS'+'\\t'\\\n",
    "                     '.\\tGT:CNADJ\\t'+str(row['data'])[:4]+str(1)+'\\t'+'0|0:0\\n'\n",
    "            \n",
    "            file_vcf.write(line)\n",
    "            count+=1\n",
    "    file_vcf.close()\n",
    "    print(\"done\")\n",
    "    \n",
    "input_file_path = \"strelka.vcf.gz\"\n",
    "output_file_path = \"sample.vcf\"\n",
    "df_snv = read_vcf_without_headers(input_file_path)\n",
    "strelka_to_sctusv(df_snv,output_file_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec3f70e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
