{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a5e3a04a-2559-44f1-9eb3-0b0f4b0b2da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "from io import StringIO\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "\n",
    "# TODO: It needs to take in Strelka SNVs, Strelka SVs and Manta SVs, CNVKit CNAs. \n",
    "\n",
    "folder_path = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling'\n",
    "\n",
    "# Take in a list of sample identifiers as a .txt within that variant folder, with an option to auto-generate based on what's in the first folder\n",
    "\n",
    "\n",
    "def convert_chromosome(chrom):\n",
    "    if chrom == 'chrX':\n",
    "        return 23\n",
    "    elif chrom == \"chrY\":\n",
    "        return 24\n",
    "    else:\n",
    "        return int(chrom[3:])\n",
    "\n",
    "\n",
    "def read_sample_list(path):\n",
    "    \"\"\"\n",
    "    Returns a list of samples to compile across Sarek output that has different Strelka, Mutect, Manta and CNVKit output in separate folders.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path: a path to a .txt file with the appropriate common folder headers. \n",
    "    \"\"\"\n",
    "    pass\n",
    "\n",
    "# Function to init the vcf file with the sample name. \n",
    "\n",
    "def init_vcf(output_folder, sample):\n",
    "\n",
    "    \"\"\"\n",
    "    Initializes the vcf output in the appropriate folder,\n",
    "\n",
    "    Parameters:\n",
    "    -------------\n",
    "\n",
    "    output_folder: folder that will have the per-sample.vcf files compiled from various callers, this will be the input for TUSV-ext.\n",
    "    sample: sample name, which will become the filename of the vcf file.\n",
    "    \n",
    "    \"\"\"\n",
    "    output_file_path = os.path.join(output_folder, sample + '.vcf')\n",
    "    \n",
    "    file_vcf = open(output_file_path,'w')\n",
    "    file_vcf.write('##fileformat=VCFv4.2\\n\\\n",
    "##filedate=20211011\\n\\\n",
    "##INFO=<ID=END,Number=1,Type=Integer,Description=\"End position of the variant described in this record\">\\n\\\n",
    "##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description=\"Imprecise structural variation\">\\n\\\n",
    "##INFO=<ID=MATEID,Number=.,Type=String,Description=\"ID of mate breakends\">\\n\\\n",
    "##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">\\n\\\n",
    "##FORMAT=<ID=GT,Number=1,Type=Integer,Description=\"Genotype\">\\n\\\n",
    "##FORMAT=<ID=CN,Number=2,Type=Integer,Description=\"Copy number genotype for imprecise events\">\\n\\\n",
    "##FORMAT=<ID=CNADJ,Number=.,Type=Integer,Description=\"Copy number of adjacency\">\\n\\\n",
    "##FORMAT=<ID=BDP,Number=1,Type=Integer,Description=\"Depth of split reads\">\\n\\\n",
    "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read depth\">\\n\\\n",
    "##ALT=<ID=DEL,Description=\"Deletion\">\\n\\\n",
    "##ALT=<ID=DUP,Description=\"Duplication\">\\n\\\n",
    "##ALT=<ID=INS,Description=\"Insertion of novel sequence\">\\n\\\n",
    "##ALT=<ID=CNV,Description=\"Copy number variable region\">\\n\\\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tTUMOR\tNORMAL\\n')\n",
    "\n",
    "# Function to take in CNA from CNVkit\n",
    "\n",
    "def write_df_to_vcf(df, vcf_file):\n",
    "    \"\"\"\n",
    "    Writes a pandas DataFrame to a VCF file.\n",
    "    \n",
    "    Args:\n",
    "    - df (pd.DataFrame): DataFrame containing the variant data.\n",
    "    - vcf_file (str): Path to the output VCF file.\n",
    "    \"\"\"\n",
    "\n",
    "    # Write the VCF header to the file\n",
    "    with open(vcf_file, 'w') as f:\n",
    "        for line in vcf_header:\n",
    "            f.write(line + '\\n')\n",
    "    \n",
    "    # Write the DataFrame to the VCF file\n",
    "    df.to_csv(vcf_file, sep='\\t', mode='a', index=False, header=False)\n",
    "\n",
    "def load_cnvkit_cns(file_path):\n",
    "    \"\"\"\n",
    "    Reads a CNVKit .cns file and loads it into a pandas DataFrame.\n",
    "    \n",
    "    Args:\n",
    "    - file_path (str): Path to the CNVKit .cns file.\n",
    "    \n",
    "    Returns:\n",
    "    - pd.DataFrame: DataFrame containing the CNVKit .cns data.\n",
    "    \"\"\"\n",
    "    # Read the .cns file into a DataFrame\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    \n",
    "    # Ensure that the columns we need are present\n",
    "    required_columns = ['chromosome', 'start', 'end', 'log2', 'depth']\n",
    "    for col in required_columns:\n",
    "        if col not in df.columns:\n",
    "            raise ValueError(f\"Required column '{col}' not found in the .cns file.\")\n",
    "    \n",
    "    # Return the DataFrame\n",
    "    return df\n",
    "\n",
    "def cnvkit_to_df(cnvkit_df):\n",
    "    \"\"\"\n",
    "    Converts a CNVKit DataFrame into a VCF-like DataFrame.\n",
    "    \n",
    "    Args:\n",
    "    - cnvkit_df (pd.DataFrame): DataFrame containing CNVKit data.\n",
    "    \n",
    "    Returns:\n",
    "    - pd.DataFrame: DataFrame formatted for VCF output.\n",
    "    \"\"\"\n",
    "    vcf_records = []\n",
    "    for idx, row in cnvkit_df.iterrows():\n",
    "        chrom = row['chromosome']\n",
    "        pos = row['start']\n",
    "        id = f'cnv{str(idx).zfill(2)}'\n",
    "        ref = '.'\n",
    "        alt = '<CNV>'\n",
    "        qual = '.'\n",
    "        filter = 'PASS'\n",
    "        end = row['end']\n",
    "        info = f'END={end};IMPRECISE'\n",
    "        format_field = 'GT:CN'\n",
    "        tumor_cn = f'1|1:{row[\"log2\"]},{row[\"depth\"]}'\n",
    "        normal_cn = '0|0:1,1'\n",
    "        record = {\n",
    "            '#CHROM': chrom,\n",
    "            'POS': pos,\n",
    "            'ID': id,\n",
    "            'REF': ref,\n",
    "            'ALT': alt,\n",
    "            'QUAL': qual,\n",
    "            'FILTER': filter,\n",
    "            'INFO': info,\n",
    "            'FORMAT': format_field,\n",
    "            'TUMOR': tumor_cn,\n",
    "            'NORMAL': normal_cn\n",
    "        }\n",
    "        vcf_records.append(record)\n",
    "    \n",
    "    vcf_df = pd.DataFrame(vcf_records)\n",
    "    \n",
    "    return vcf_df\n",
    "\n",
    "\n",
    "# Function to take in SNVs from Strelka\n",
    "\n",
    "def read_strelka(input_file_path):\n",
    "    \"\"\"\n",
    "    Function that reads in the file path to the Strelka SNV output, processes it and returns it as a df_SNV.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    input_file_path = input file for the strelka_snv\n",
    "    \n",
    "    \"\"\"\n",
    "    filtered_lines = []\n",
    "    with gzip.open(input_file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('##'):\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "    filtered_content = ''.join(filtered_lines)\n",
    "    df_snv = pd.read_csv(StringIO(filtered_content), sep='\\t')\n",
    "    df_snv['chr'] = df_snv['#CHROM'].apply(convert_chromosome)\n",
    "    df_snv['data'] = df_snv[list(df_snv.columns)[-2]]\n",
    "\n",
    "    return df_snv\n",
    "\n",
    "def convert_strelka_to_tusv_ext_df(strelka_df):\n",
    "    # Initialize an empty list to store the rows of the new DataFrame\n",
    "    tusv_ext_rows = []\n",
    "\n",
    "    for index, row in strelka_df.iterrows():\n",
    "        chrom = row['#CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        normal_data = row['NORMAL']\n",
    "        tumor_data = row['TUMOR']\n",
    "        \n",
    "        # Example ID and other static fields for simplicity\n",
    "        variant_id = f\"snv{index:04d}\"\n",
    "        qual = \".\"\n",
    "        filter_val = \"PASS\"\n",
    "        info = \".\"\n",
    "        format_val = \"GT:CNADJ\"\n",
    "        \n",
    "        # Extract GT and CNADJ values\n",
    "        normal_gt = \"0|0\"\n",
    "        tumor_gt = \"0|1\"\n",
    "        normal_cnadj = \"0\"\n",
    "        tumor_cnadj = \"0\"\n",
    "        \n",
    "        normal = f\"{normal_gt}:{normal_cnadj}\"\n",
    "        tumor = f\"{tumor_gt}:{tumor_cnadj}\"\n",
    "        \n",
    "        tusv_ext_rows.append([\n",
    "            chrom, pos, variant_id, ref, alt, qual, filter_val, info, format_val, tumor, normal\n",
    "        ])\n",
    "    \n",
    "    # Create a new DataFrame with the appropriate columns\n",
    "    tusv_ext_df = pd.DataFrame(tusv_ext_rows, columns=[\n",
    "        '#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'TUMOR', 'NORMAL'\n",
    "    ])\n",
    "    \n",
    "    return tusv_ext_df\n",
    "\n",
    "\n",
    "\n",
    "# Function to take in large SVs from Manta\n",
    "\n",
    "def read_manta_sv(input_file_path):\n",
    "\n",
    "\n",
    "    filtered_lines = []\n",
    "    with gzip.open(input_file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('##'):\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "    filtered_content = ''.join(filtered_lines)\n",
    "    df_manta_sv = pd.read_csv(StringIO(filtered_content), sep='\\t')\n",
    "\n",
    "    \n",
    "    df_processed_manta = rewrite_manta_results(df_manta_sv)\n",
    "    \n",
    "    return df_processed_manta\n",
    "\n",
    "\n",
    "def rewrite_manta_results(df):\n",
    "    new_rows = []\n",
    "\n",
    "    print(df.shape)\n",
    "    df = df[df[\"FILTER\"] == \"PASS\"]\n",
    "    print(df.shape)\n",
    "    \n",
    "    # Identify TUMOR and NORMAL columns\n",
    "    tumor_col = next(col for col in df.columns if col.split('-')[-1].startswith('T'))\n",
    "    normal_col = next(col for col in df.columns if col.split('-')[-1].startswith('N'))\n",
    "    \n",
    "    id_counter = 0\n",
    "\n",
    "    for _, row in df.iterrows():\n",
    "        id_counter += 1\n",
    "        sv_id = f\"sv{id_counter:02d}\"\n",
    "        \n",
    "        info_dict = dict(item.split('=') for item in row['INFO'].split(';') if '=' in item)\n",
    "        mate_id = info_dict.get('MATEID', '').split(':')[0]\n",
    "        sv_type = info_dict.get('SVTYPE', '')\n",
    "        sv_len = info_dict.get('SVLEN', '')\n",
    "        end = info_dict.get('END', '')\n",
    "        \n",
    "        alt = row['ALT']\n",
    "        if sv_type == 'BND':\n",
    "            match = re.search(r'\\[(.*?)\\[|\\](.*?)\\]', alt)\n",
    "            if match:\n",
    "                chr_pos = match.group(1) or match.group(2)\n",
    "                chr, pos = chr_pos.split(':')\n",
    "                alt = f\"[{chr}:{pos}[\" if '[' in row['ALT'] else f\"]{chr}:{pos}]\"\n",
    "        \n",
    "        # Extract PR and SR from FORMAT field\n",
    "        format_fields = row['FORMAT'].split(':')\n",
    "        tumor_values = row[tumor_col].split(':')\n",
    "        normal_values = row[normal_col].split(':')\n",
    "        \n",
    "        format_dict = dict(zip(format_fields, tumor_values))\n",
    "        pr_tumor = format_dict.get('PR', '0,0')\n",
    "        sr_tumor = format_dict.get('SR', '0,0')\n",
    "        \n",
    "        format_dict = dict(zip(format_fields, normal_values))\n",
    "        pr_normal = format_dict.get('PR', '0,0')\n",
    "        sr_normal = format_dict.get('SR', '0,0')\n",
    "        \n",
    "        new_row = {\n",
    "            '#CHROM': row['#CHROM'],\n",
    "            'POS': row['POS'],\n",
    "            'ID': sv_id,\n",
    "            'REF': row['REF'],\n",
    "            'ALT': alt,\n",
    "            'QUAL': row['QUAL'],\n",
    "            'FILTER': row['FILTER'],\n",
    "            'INFO': f\"MATEID={mate_id};SVTYPE={sv_type};SVLEN={sv_len};END={end}\",\n",
    "            'FORMAT': 'GT:PR:SR',\n",
    "            'TUMOR': f\"{tumor_values[0]}:{pr_tumor}:{sr_tumor}\",\n",
    "            'NORMAL': f\"{normal_values[0]}:{pr_normal}:{sr_normal}\"\n",
    "        }\n",
    "        \n",
    "        new_rows.append(new_row)\n",
    "    \n",
    "    new_df = pd.DataFrame(new_rows, columns=['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'TUMOR', 'NORMAL'])\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "174cd4e2-fb66-485b-b9b3-fcafa3ced517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #CHROM    POS ID REF ALT QUAL  FILTER  \\\n",
      "0   chr1  10120  .   T   C    .  LowEVS   \n",
      "1   chr1  10257  .   A   C    .  LowEVS   \n",
      "2   chr1  10330  .   C   A    .  LowEVS   \n",
      "3   chr1  14063  .   G   A    .  LowEVS   \n",
      "4   chr1  14653  .   C   T    .  LowEVS   \n",
      "\n",
      "                                                INFO  \\\n",
      "0  DP=1080;MQ=12.08;MQ0=874;NT=ref;QSS=8;QSS_NT=8...   \n",
      "1  DP=190;MQ=16.38;MQ0=92;NT=ref;QSS=3;QSS_NT=3;R...   \n",
      "2  DP=212;MQ=21.15;MQ0=89;NT=ref;QSS=1;QSS_NT=1;R...   \n",
      "3  DP=37;MQ=18.01;MQ0=25;NT=ref;QSS=1;QSS_NT=1;Re...   \n",
      "4  DP=45;MQ=29.11;MQ0=14;NT=ref;QSS=4;QSS_NT=4;Re...   \n",
      "\n",
      "                         FORMAT                         NORMAL  \\\n",
      "0  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  48:18:1:0:2,19:0,2:0,2:28,112   \n",
      "1  DP:FDP:SDP:SUBDP:AU:CU:GU:TU      12:7:1:0:5,41:0,5:0,0:0,0   \n",
      "2  DP:FDP:SDP:SUBDP:AU:CU:GU:TU     26:19:6:0:0,5:7,23:0,1:0,1   \n",
      "3  DP:FDP:SDP:SUBDP:AU:CU:GU:TU       3:0:0:0:0,0:0,0:3,14:0,0   \n",
      "4  DP:FDP:SDP:SUBDP:AU:CU:GU:TU     14:0:0:0:0,0:12,18:0,0:2,3   \n",
      "\n",
      "                          TUMOR  chr                          data  \n",
      "0  48:17:0:0:2,14:3,4:0,5:26,91    1  48:17:0:0:2,14:3,4:0,5:26,91  \n",
      "1   19:14:2:0:3,50:2,10:0,0:0,0    1   19:14:2:0:3,50:2,10:0,0:0,0  \n",
      "2   29:22:1:0:2,11:5,24:0,0:0,0    1   29:22:1:0:2,11:5,24:0,0:0,0  \n",
      "3      4:0:0:0:2,2:0,0:2,21:0,0    1      4:0:0:0:2,2:0,0:2,21:0,0  \n",
      "4    15:0:0:0:0,0:13,22:0,0:2,2    1    15:0:0:0:0,0:13,22:0,0:2,2  \n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in Strelka indel files to test\n",
    "\n",
    "test_strelka_indel_path = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/strelka/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.strelka.somatic_snvs.vcf.gz'\n",
    "\n",
    "df_indel = read_strelka(test_strelka_snv_path)\n",
    "\n",
    "\n",
    "print(df_indel.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4e3464e2-58c7-47ed-bd5c-2f7d79896fec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #CHROM    POS       ID REF ALT QUAL FILTER INFO    FORMAT  TUMOR NORMAL\n",
      "0   chr1  10120  snv0000   T   C    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "1   chr1  10257  snv0001   A   C    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "2   chr1  10330  snv0002   C   A    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "3   chr1  14063  snv0003   G   A    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "4   chr1  14653  snv0004   C   T    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in Strelka snv files to test.\n",
    "\n",
    "\n",
    "test_strelka_snv_path = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/strelka/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.strelka.somatic_indels.vcf.gz'\n",
    "\n",
    "df_snv = read_strelka_snv(test_strelka_snv_path)\n",
    "\n",
    "tusv_ext_snv = convert_strelka_to_tusv_ext_df(df_snv)\n",
    "print(tusv_ext_snv.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f9a2a282-781c-4c9e-9d2b-c22101d9440e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chromosome    start       end gene      log2     depth  probes     weight  \\\n",
      "0       chr1    14034    351968    -  0.098452  14.55930      58    52.1960   \n",
      "1       chr1   351968    459979    - -0.413539   7.87716      27    19.1118   \n",
      "2       chr1   459979   1264807    - -0.022918  13.51170     180   170.4730   \n",
      "3       chr1  1264807   1333091    -  0.322362  17.72340      17    16.8359   \n",
      "4       chr1  1333091  16558479    - -0.011359  13.51330    3752  3682.8400   \n",
      "\n",
      "      ci_lo     ci_hi  \n",
      "0  0.080594  0.119033  \n",
      "1 -0.444376 -0.386697  \n",
      "2 -0.031888 -0.010873  \n",
      "3  0.294485  0.344509  \n",
      "4 -0.013140 -0.009817  \n",
      "  #CHROM      POS     ID REF    ALT QUAL FILTER                    INFO  \\\n",
      "0   chr1    14034  cnv00   .  <CNV>    .   PASS    END=351968;IMPRECISE   \n",
      "1   chr1   351968  cnv01   .  <CNV>    .   PASS    END=459979;IMPRECISE   \n",
      "2   chr1   459979  cnv02   .  <CNV>    .   PASS   END=1264807;IMPRECISE   \n",
      "3   chr1  1264807  cnv03   .  <CNV>    .   PASS   END=1333091;IMPRECISE   \n",
      "4   chr1  1333091  cnv04   .  <CNV>    .   PASS  END=16558479;IMPRECISE   \n",
      "\n",
      "  FORMAT                   TUMOR   NORMAL  \n",
      "0  GT:CN   1|1:0.0984518,14.5593  0|0:1,1  \n",
      "1  GT:CN   1|1:-0.413539,7.87716  0|0:1,1  \n",
      "2  GT:CN  1|1:-0.0229176,13.5117  0|0:1,1  \n",
      "3  GT:CN    1|1:0.322362,17.7234  0|0:1,1  \n",
      "4  GT:CN  1|1:-0.0113593,13.5133  0|0:1,1  \n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in CNVKit .cns files to test.\n",
    "\n",
    "test_cnvkit_cns_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/cnvkit/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4.cns'\n",
    "\n",
    "cnvkit_df = load_cnvkit_cns(test_cnvkit_cns_file)\n",
    "print(cnvkit_df.head())\n",
    "\n",
    "vcf_df = cnvkit_to_df(cnvkit_df)\n",
    "print(vcf_df.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e89c704e-7841-4e28-bfba-2e5466d8d68a",
   "metadata": {},
   "outputs": [
    {
     "ename": "StopIteration",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m test_manta_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/manta/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.manta.somatic_sv.vcf.gz\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      4\u001b[0m test_organoid_wes_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/nfcore/achang4_bridgesPSC/sarek/Part2_FASTQ2SAREK/results/variant_calling/manta/123_organoid_vs_123_blood/123_organoid_vs_123_blood.manta.somatic_sv.vcf.gz\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 5\u001b[0m df_processed_manta \u001b[38;5;241m=\u001b[39m \u001b[43mread_manta_sv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtest_organoid_wes_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28mprint\u001b[39m(df_processed_manta\u001b[38;5;241m.\u001b[39mhead())\n",
      "Cell \u001b[0;32mIn[25], line 236\u001b[0m, in \u001b[0;36mread_manta_sv\u001b[0;34m(input_file_path)\u001b[0m\n\u001b[1;32m    232\u001b[0m filtered_content \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines)\n\u001b[1;32m    233\u001b[0m df_manta_sv \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(StringIO(filtered_content), sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 236\u001b[0m df_processed_manta \u001b[38;5;241m=\u001b[39m \u001b[43mrewrite_manta_results\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf_manta_sv\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    238\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m df_processed_manta\n",
      "Cell \u001b[0;32mIn[25], line 245\u001b[0m, in \u001b[0;36mrewrite_manta_results\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m    242\u001b[0m new_rows \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    244\u001b[0m \u001b[38;5;66;03m# Identify TUMOR and NORMAL columns\u001b[39;00m\n\u001b[0;32m--> 245\u001b[0m tumor_col \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mcol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mcol\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstartswith\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mT\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    246\u001b[0m normal_col \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(col \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;28;01mif\u001b[39;00m col\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mN\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m    248\u001b[0m id_counter \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "\u001b[0;31mStopIteration\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Test box for reading in Manta .vcf files to determine format.\n",
    "\n",
    "test_manta_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/manta/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.manta.somatic_sv.vcf.gz'\n",
    "test_organoid_wes_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/nfcore/achang4_bridgesPSC/sarek/Part2_FASTQ2SAREK/results/variant_calling/manta/123_organoid_vs_123_blood/123_organoid_vs_123_blood.manta.somatic_sv.vcf.gz'\n",
    "df_processed_manta = read_manta_sv(test_organoid_wes_file)\n",
    "\n",
    "print(df_processed_manta.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c1427d-4c89-4117-8ca4-b25b6d4a540e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42f9386-5616-4fcd-b384-f3a39a03a1f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LOpaperqa",
   "language": "python",
   "name": "lo_paperqa"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
