{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "24d92fda-3a55-4145-94b0-0353a05933ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUR-AER5-TTM4_vs_AUR-AER5-NT2\n",
      "{'1': 248956422, '2': 242193529, '3': 198295559, '4': 190214555, '5': 181538259, '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, '10': 133797422, '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189, '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '20': 64444167, '21': 46709983, '22': 50818468, '23': 156040895, '24': 57227415}\n",
      "  chr   startpos     endpos  nMajor  nMinor\n",
      "0   1     873251   35918614       2       1\n",
      "1   1   36100398  125122318       2       2\n",
      "2   1  145837693  160796870       6       1\n",
      "3   1  160797135  161379107      11       1\n",
      "4   1  161509955  167683765       6       1\n",
      "5   1  167765556  248916508       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "6    2      41404  119246204       3       2\n",
      "7    2  119301482  124764368       3       1\n",
      "8    2  124789687  128797876       3       2\n",
      "9    2  129496404  192699777       3       1\n",
      "10   2  193277860  241744700       3       0\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "11   3      29609   52012813       2       1\n",
      "12   3   52075825   84610573       1       0\n",
      "13   3   84884887  195568790       2       1\n",
      "14   3  195752197  195760866       5       1\n",
      "15   3  195763273  196369580       4       1\n",
      "16   3  196391576  198058644       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "17   4     85531   69638595       1       1\n",
      "18   4  69647833   70592568       2       1\n",
      "19   4  70601197  189446500       1       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "20   5     140417    5443066       1       1\n",
      "21   5    5625424    8586789       1       0\n",
      "22   5    8893267  135850386       1       1\n",
      "23   5  135893944  170627737       2       1\n",
      "24   5  170669425  170968508       4       1\n",
      "25   5  171213458  181155980       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "26   6    397290   33086205       4       0\n",
      "27   6  33086269   33103820       2       0\n",
      "28   6  33103823   41229296       4       0\n",
      "29   6  41229696  170583760       3       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "30   7     81400   57293253       2       1\n",
      "31   7  57293271  159295340       2       2\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "32   8    244884   34666995       1       0\n",
      "33   8  34874154   52036608       3       1\n",
      "34   8  52036682  144995060       4       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "35   9    154933  138181722       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "36  10      48486   11485787       2       2\n",
      "37  10   11488873   12538590       4       2\n",
      "38  10   12538665   45965799       2       2\n",
      "39  10   46003078  131259095       2       0\n",
      "40  10  131737112  133557962       3       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "41  11    199813   50422246       3       1\n",
      "42  11  54592062  135030575       1       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "43  12    140964    6155997       1       1\n",
      "44  12   6235434    6982757       2       1\n",
      "45  12   7064937   72621882       1       1\n",
      "46  12  72679179  133219447       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "47  13  18854918  114324544       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "48  14  19921447   23118022       2       0\n",
      "49  14  23127111   23640185       4       1\n",
      "50  14  23640586   57409547       2       0\n",
      "51  14  57478656   58211551       1       0\n",
      "52  14  58258099  106681273       2       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "53  15  19897077   88144866       2       0\n",
      "54  15  88334746   89136284       6       0\n",
      "55  15  89176144  101829176       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "56  16     34170  90096325       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "57  17    161952  39984674       2       0\n",
      "58  17  39986424  40300899       4       0\n",
      "59  17  40436738  43051308       2       0\n",
      "60  17  43063808  71563724       3       0\n",
      "61  17  71621433  76226239       4       0\n",
      "62  17  76226409  83048974       3       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "63  18     86837  34793772       3       2\n",
      "64  18  34793909  80159596       2       2\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "65  19    282753  31592317       2       2\n",
      "66  19  31958935  58478128       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "67  20    145888  18041527       3       1\n",
      "68  20  18138657  38230933       2       1\n",
      "69  20  38293455  38318446       6       1\n",
      "70  20  38324695  38434831       4       1\n",
      "71  20  38517732  64273189       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "72  21  13249222  46602462       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "73  22  16104906  50580150       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "74  23    346920  155774775       2       0\n",
      "['PR:SR']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/envs/tusv_ext/lib/python3.6/site-packages/ipykernel_launcher.py:415: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUR-AER5-TTM5_vs_AUR-AER5-NT2\n",
      "{'1': 248956422, '2': 242193529, '3': 198295559, '4': 190214555, '5': 181538259, '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, '10': 133797422, '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189, '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '20': 64444167, '21': 46709983, '22': 50818468, '23': 156040895, '24': 57227415}\n",
      "  chr   startpos     endpos  nMajor  nMinor\n",
      "0   1     873251   35918614       2       1\n",
      "1   1   36100398  125122318       2       2\n",
      "2   1  145837693  160670822       4       1\n",
      "3   1  160796868  161379107       8       1\n",
      "4   1  161509955  164935503       4       1\n",
      "5   1  165204312  167854735       3       1\n",
      "6   1  167856247  248916508       2       1\n",
      "  chr   startpos     endpos  nMajor  nMinor\n",
      "7   2      41404   70165247       3       2\n",
      "8   2   70165688  103825205       4       2\n",
      "9   2  103825227  241744700       4       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "10   3      29609   10219564       1       1\n",
      "11   3   10219610   10498393       2       1\n",
      "12   3   10498638   52012813       1       1\n",
      "13   3   52075825   84610573       1       0\n",
      "14   3   84884887  195568790       2       1\n",
      "15   3  195752197  195760866       5       1\n",
      "16   3  195763273  196391576       4       1\n",
      "17   3  196391750  198058644       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "18   4     85531  189446500       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "19   5     140417    5443066       1       1\n",
      "20   5    5625424    8549865       1       0\n",
      "21   5    8893267   45830406       1       1\n",
      "22   5   45830508  135850386       2       1\n",
      "23   5  135893944  178261771       2       2\n",
      "24   5  178263538  181155980       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "25   6     397290  136791495       3       0\n",
      "26   6  136866468  170583760       4       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "27   7     81400   57293253       2       1\n",
      "28   7  57293271  159295340       2       2\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "29   8    244884   34666995       1       0\n",
      "30   8  34874154   34934598       4       0\n",
      "31   8  34934629   37095523       3       0\n",
      "32   8  37096128   37187658       1       0\n",
      "33   8  37361134   52131787       3       1\n",
      "34   8  52136431  144995060       4       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "35   9     154933  129084495       2       1\n",
      "36   9  129086103  138181722       2       2\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "37  10      48486   11485787       2       1\n",
      "38  10   11488873   12785661       4       1\n",
      "39  10   12785691   45965799       2       1\n",
      "40  10   46003078  131259095       2       0\n",
      "41  10  131737112  133557962       3       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "42  11    199813   50422246       3       1\n",
      "43  11  54592062  135030575       1       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "44  12    140964    6235434       3       0\n",
      "45  12   6235914    6982757       5       0\n",
      "46  12   7064937   34166374       3       0\n",
      "47  12  37454374  133219447       2       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "48  13  18854918  114324544       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "49  14  19921447   23118022       2       0\n",
      "50  14  23127111   23776750       5       1\n",
      "51  14  23776890   48724565       2       0\n",
      "52  14  48995584   50482668       4       0\n",
      "53  14  50587880  106681273       2       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "54  15  19897077   88144866       2       0\n",
      "55  15  88334746   89136284       6       0\n",
      "56  15  89176144  101829176       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "57  16     34170  90096325       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "58  17    161952   9576047       3       0\n",
      "59  17   9584120  20754335       2       0\n",
      "60  17  20754485  27019666       3       0\n",
      "61  17  27047846  39653965       2       0\n",
      "62  17  39673351  39863888       3       0\n",
      "63  17  39871147  40300899       4       0\n",
      "64  17  40436738  43051308       2       0\n",
      "65  17  43063808  71563724       3       0\n",
      "66  17  71621433  76167409       4       0\n",
      "67  17  76226239  83048974       3       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "68  18     86837  33748555       3       2\n",
      "69  18  33748756  49992047       2       2\n",
      "70  18  49992797  80159596       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "71  19    282753   2249478       2       1\n",
      "72  19   2250529  31592317       2       2\n",
      "73  19  31958935  58478128       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "74  20    145888  18041527       3       1\n",
      "75  20  18138657  38212891       2       1\n",
      "76  20  38230933  38318446       6       1\n",
      "77  20  38324695  38434831       5       1\n",
      "78  20  38517732  62943308       2       1\n",
      "79  20  62944604  64273189       1       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "80  21  13249222  46602462       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "81  22  16104906  50580150       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "82  23    346920  155774775       2       0\n",
      "['PR:SR' 'PR']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/envs/tusv_ext/lib/python3.6/site-packages/ipykernel_launcher.py:415: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUR-AER5-TTP2_vs_AUR-AER5-NT2\n",
      "{'1': 248956422, '2': 242193529, '3': 198295559, '4': 190214555, '5': 181538259, '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, '10': 133797422, '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189, '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '20': 64444167, '21': 46709983, '22': 50818468, '23': 156040895, '24': 57227415}\n",
      "  chr   startpos     endpos  nMajor  nMinor\n",
      "0   1     873251   35918614       3       0\n",
      "1   1   36100398   36286551       4       1\n",
      "2   1   36290034   60180520       4       0\n",
      "3   1   60476692  125122318       4       1\n",
      "4   1  145837693  160670822       3       3\n",
      "5   1  160796868  161379107       6       3\n",
      "6   1  161509955  167854274       3       3\n",
      "7   1  167854734  248916508       3       2\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "8    2      41404   19102061       1       0\n",
      "9    2   19347034  164493392       2       1\n",
      "10   2  164494698  166046718       2       0\n",
      "11   2  166046935  241744700       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "12   3      29609   52012813       2       2\n",
      "13   3   52075825   84610573       2       0\n",
      "14   3   84884887  160568858       2       2\n",
      "15   3  160694366  161922700       5       2\n",
      "16   3  162237689  198058644       2       2\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "17   4      85531   69638595       2       1\n",
      "18   4   69647833   70604457       4       1\n",
      "19   4   70636027  174139617       2       1\n",
      "20   4  174239336  176301927       3       1\n",
      "21   4  176301994  189446500       1       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "22   5     140417   50263913       5       1\n",
      "23   5   50263930  120817787       1       1\n",
      "24   5  120817924  181155980       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "25   6    397290  170583760       2       2\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "26   7     81400    8263779       2       1\n",
      "27   7   8278509  159295340       2       2\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "28   8     244884   34666995       1       0\n",
      "29   8   34874154   34934598       6       1\n",
      "30   8   34934629   37095523       3       1\n",
      "31   8   37096128   37187658       2       0\n",
      "32   8   37361134  142424376       6       1\n",
      "33   8  142481726  144995060       7       3\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "34   9     154933   42919752       3       2\n",
      "35   9   63728648  133162584       3       1\n",
      "36   9  133162643  133939089       3       0\n",
      "37   9  134079343  138181722       3       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "38  10      48486   35074049       2       2\n",
      "39  10   35178743  110590121       2       0\n",
      "40  10  110669121  110986008       5       0\n",
      "41  10  111203394  133557962       2       0\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "42  11     199813   62062697       2       1\n",
      "43  11   62128743   62598193       4       1\n",
      "44  11   62627409   98565802       2       1\n",
      "45  11   98565944  102794938       2       2\n",
      "46  11  102796332  135030575       2       1\n",
      "   chr   startpos     endpos  nMajor  nMinor\n",
      "47  12     140964  130003380       2       1\n",
      "48  12  130003581  133219447       2       2\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "49  13  18854918   95411600       1       1\n",
      "50  13  95434226  114324544       2       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "51  14  19921447   23118022       2       0\n",
      "52  14  23127111   23574391       5       2\n",
      "53  14  23640185   48724565       3       0\n",
      "54  14  48995584   50083524       5       0\n",
      "55  14  50106856   50130294       4       0\n",
      "56  14  50180772   50377401       5       0\n",
      "57  14  50378138   50482668       4       0\n",
      "58  14  50587880  106681273       3       0\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "59  15  19897077   51210647       2       0\n",
      "60  15  51210789   51449382       4       0\n",
      "61  15  51456413   88144866       2       0\n",
      "62  15  88334746   89116194       6       0\n",
      "63  15  89116220  101829176       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "64  16     34170  90096325       2       1\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "65  17    161952  22717231       2       0\n",
      "66  17  26764148  40300899       4       0\n",
      "67  17  40436738  70634845       3       0\n",
      "68  17  70942026  76332605       5       0\n",
      "69  17  76332623  83048974       3       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "70  18     86837  22993903       2       2\n",
      "71  18  22997473  80159596       2       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "72  19    282753  20542844       2       1\n",
      "73  19  20544327  31592317       3       2\n",
      "74  19  31958935  58478128       3       0\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "75  20    145888  38304396       2       1\n",
      "76  20  38308738  38360727       4       2\n",
      "77  20  38363737  64273189       2       2\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "78  21  13249222  34518111       2       2\n",
      "79  21  34521830  44089769       2       1\n",
      "80  21  44093927  46602462       2       2\n",
      "   chr  startpos    endpos  nMajor  nMinor\n",
      "81  22  16104906  50580150       3       1\n",
      "   chr  startpos     endpos  nMajor  nMinor\n",
      "82  23    346920  155774775       2       0\n",
      "['PR:SR']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/envs/tusv_ext/lib/python3.6/site-packages/ipykernel_launcher.py:415: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# ChatGPT modified version 2.0 - all methods have to be carefully reviewed.\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "from io import StringIO\n",
    "import re\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# Define global variables\n",
    "cnv_count = 1\n",
    "sv_count = 1\n",
    "snv_count = 1\n",
    "\n",
    "def increment_cnv():\n",
    "    global cnv_count\n",
    "    cnv_count += 1\n",
    "\n",
    "def increment_sv():\n",
    "    global sv_count\n",
    "    sv_count += 1\n",
    "\n",
    "def increment_snv():\n",
    "    global snv_count\n",
    "    snv_count += 1\n",
    "\n",
    "def convert_chromosome(chrom):\n",
    "    if chrom == 'chrX' or chrom == 'X':\n",
    "        return \"23\"\n",
    "    elif chrom == \"chrY\" or chrom == 'Y':\n",
    "        return \"24\"\n",
    "    elif chrom.startswith('chr'):\n",
    "        return str(chrom[3:])\n",
    "    else:\n",
    "        return str(chrom)\n",
    "\n",
    "def init_vcf(output_folder, sample):\n",
    "    output_file_path = os.path.join(output_folder, sample + '.vcf')\n",
    "    with open(output_file_path, 'w') as file_vcf:\n",
    "        file_vcf.write('##fileformat=VCFv4.2\\n')\n",
    "        file_vcf.write('##filedate=20211011\\n')\n",
    "        file_vcf.write('##INFO=<ID=END,Number=1,Type=Integer,Description=\"End position of the variant described in this record\">\\n')\n",
    "        file_vcf.write('##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description=\"Imprecise structural variation\">\\n')\n",
    "        file_vcf.write('##INFO=<ID=MATEID,Number=.,Type=String,Description=\"ID of mate breakends\">\\n')\n",
    "        file_vcf.write('##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">\\n')\n",
    "        file_vcf.write('##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\\n')\n",
    "        file_vcf.write('##FORMAT=<ID=CN,Number=2,Type=Integer,Description=\"Copy number genotype for imprecise events\">\\n')\n",
    "        file_vcf.write('##FORMAT=<ID=CNADJ,Number=.,Type=Float,Description=\"Copy number of adjacency\">\\n')\n",
    "        file_vcf.write('##FORMAT=<ID=BDP,Number=1,Type=Integer,Description=\"Depth of split reads\">\\n')\n",
    "        file_vcf.write('##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read depth\">\\n')\n",
    "        file_vcf.write('##ALT=<ID=DEL,Description=\"Deletion\">\\n')\n",
    "        file_vcf.write('##ALT=<ID=DUP,Description=\"Duplication\">\\n')\n",
    "        file_vcf.write('##ALT=<ID=INS,Description=\"Insertion of novel sequence\">\\n')\n",
    "        file_vcf.write('##ALT=<ID=CNV,Description=\"Copy number variable region\">\\n')\n",
    "        file_vcf.write('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tTUMOR\\tNORMAL\\n')\n",
    "    return output_file_path\n",
    "\n",
    "def write_df_to_vcf(df, vcf_file):\n",
    "    # Filter out rows where #CHROM column has the value 24\n",
    "    # Ensure the #CHROM column is treated as a string\n",
    "    df['#CHROM'] = df['#CHROM'].astype(str)\n",
    "    \n",
    "    # Perform the filtering operation\n",
    "    filtered_df = df[df['#CHROM'] != '24']\n",
    "    filtered_df.to_csv(vcf_file, sep='\\t', mode='a', index=False, header=False)\n",
    "\n",
    "\n",
    "def load_ascat_cnv(file_path):\n",
    "    \"\"\"\n",
    "    Reads an ASCAT output file into a pandas DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    file_path (str): The path to the ASCAT output file.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the ASCAT data.\n",
    "    \"\"\"\n",
    "    # Define column names as specified in the example\n",
    "    column_names = ['chr', 'startpos', 'endpos', 'nMajor', 'nMinor']\n",
    "    \n",
    "    # Read the file into a DataFrame\n",
    "    df = pd.read_csv(file_path, sep='\\t', names=column_names, header=0)\n",
    "    df['chr'] = df['chr'].apply(convert_chromosome)\n",
    "    return df\n",
    "\n",
    "def convert_ascat_to_vcf_df(df, reference_genome_path):\n",
    "    \"\"\"\n",
    "    Converts a DataFrame with ASCAT data to a VCF compatible DataFrame,\n",
    "    adding rows for all segments of each chromosome that are not already in the data.\n",
    "    \"\"\"\n",
    "    vcf_records = []\n",
    "    chromosome_lengths = {}\n",
    "    cnv_count = 0\n",
    "\n",
    "\n",
    "\n",
    "    # Parse the reference genome file\n",
    "    for record in SeqIO.parse(reference_genome_path, \"fasta\"):\n",
    "        chromosome = record.id\n",
    "        length = len(record.seq)\n",
    "        if chromosome.startswith(\"NC_0000\"):\n",
    "            chromosome_number = chromosome.split('.')[0].replace(\"NC_0000\", \"\")\n",
    "            chromosome_number = str(int(chromosome_number))  # Remove leading zeros\n",
    "            chromosome_lengths[chromosome_number] = length\n",
    "        elif chromosome.startswith(\"NC_012920.1\"):\n",
    "            # Skip the MT chromosome\n",
    "            continue\n",
    "\n",
    "    # Sort chromosome lengths based on chromosome number\n",
    "    sorted_chromosome_lengths = dict(sorted(chromosome_lengths.items(), key=lambda item: int(item[0])))\n",
    "    print(sorted_chromosome_lengths)\n",
    "    # Process each chromosome\n",
    "    for chrom, length in sorted_chromosome_lengths.items():\n",
    "        if chrom == \"24\" and \"24\" not in df['chr'].values:\n",
    "            continue\n",
    "\n",
    "        # Map chromosome \"23\" to \"X\" and \"24\" to \"Y\" for data lookup\n",
    "        lookup_chrom = \"23\" if chrom == \"23\" else (\"24\" if chrom == \"24\" else chrom)\n",
    "        \n",
    "        chrom_df = df[df['chr'] == str(lookup_chrom)]\n",
    "        sorted_chrom_df = chrom_df.sort_values(by='startpos')\n",
    "        \n",
    "        last_end = 0\n",
    "        print(sorted_chrom_df)\n",
    "        # If no data for this chromosome, add a single entry covering the entire chromosome\n",
    "        if sorted_chrom_df.empty:\n",
    "            record_id = increment_cnv()\n",
    "            info = f\"END={length};IMPRECISE\"\n",
    "            format_field = \"GT:CN\"\n",
    "            tumor_sample = \"1|1:1,1\"\n",
    "            normal_sample = \"0|0:1,1\"\n",
    "            vcf_records.append([chrom, 1, record_id, \".\", \"<CNV>\", \".\", \"PASS\", info, format_field, tumor_sample, normal_sample])\n",
    "        else:\n",
    "            for idx, row in sorted_chrom_df.iterrows():\n",
    "                start = row['startpos']\n",
    "                end = row['endpos']\n",
    "                nMajor = row['nMajor']\n",
    "                nMinor = row['nMinor']\n",
    "\n",
    "                # Fill gap before this segment if needed\n",
    "                if start > last_end + 1:\n",
    "                    record_id = increment_cnv()\n",
    "                    info = f\"END={start-1};IMPRECISE\"\n",
    "                    format_field = \"GT:CN\"\n",
    "                    tumor_sample = \"1|1:1,1\"\n",
    "                    normal_sample = \"0|0:1,1\"\n",
    "                    vcf_records.append([chrom, last_end + 1, record_id, \".\", \"<CNV>\", \".\", \"PASS\", info, format_field, tumor_sample, normal_sample])\n",
    "\n",
    "                # Add the segment from ASCAT data\n",
    "                record_id = increment_cnv()\n",
    "                info = f\"END={end};IMPRECISE\"\n",
    "                format_field = \"GT:CN\"\n",
    "                tumor_sample = f\"1|1:{nMajor},{nMinor}\"\n",
    "                normal_sample = \"0|0:1,1\"\n",
    "                vcf_records.append([chrom, start, record_id, \".\", \"<CNV>\", \".\", \"PASS\", info, format_field, tumor_sample, normal_sample])\n",
    "\n",
    "                last_end = end\n",
    "\n",
    "            # Fill gap after last segment if needed\n",
    "            if last_end < length:\n",
    "                record_id = increment_cnv()\n",
    "                info = f\"END={length};IMPRECISE\"\n",
    "                format_field = \"GT:CN\"\n",
    "                tumor_sample = \"1|1:1,1\"\n",
    "                normal_sample = \"0|0:1,1\"\n",
    "                vcf_records.append([chrom, last_end + 1, record_id, \".\", \"<CNV>\", \".\", \"PASS\", info, format_field, tumor_sample, normal_sample])\n",
    "\n",
    "    # Create the VCF DataFrame\n",
    "    vcf_df = pd.DataFrame(vcf_records, columns=[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\", \"FORMAT\", \"TUMOR\", \"NORMAL\"])\n",
    "    \n",
    "    # Ensure #CHROM and POS are int columns before sorting\n",
    "    vcf_df[\"#CHROM\"] = vcf_df[\"#CHROM\"].astype(int)\n",
    "    vcf_df[\"POS\"] = vcf_df[\"POS\"].astype(int)\n",
    "    \n",
    "    # Sort the VCF DataFrame by chromosome and position\n",
    "    vcf_df = vcf_df.sort_values(by=[\"#CHROM\", \"POS\"]).reset_index(drop=True)\n",
    "    \n",
    "    # Reassign CNV IDs in numerical ascending order\n",
    "    vcf_df[\"ID\"] = [\"cnv\" + str(i + 1) for i in range(len(vcf_df))]\n",
    "    \n",
    "    return vcf_df\n",
    "\n",
    "def read_strelka(input_file_path):\n",
    "    filtered_lines = []\n",
    "    with gzip.open(input_file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('##'):\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "    filtered_content = ''.join(filtered_lines)\n",
    "    df_snv = pd.read_csv(StringIO(filtered_content), sep='\\t')\n",
    "    df_snv['#CHROM'] = df_snv['#CHROM'].apply(convert_chromosome)\n",
    "    return df_snv\n",
    "\n",
    "def get_position(input_string, search):\n",
    "    # Split the input string by ':'\n",
    "    split_string = input_string.split(':')\n",
    "    \n",
    "    # Find the position of 'AU'\n",
    "    try:\n",
    "        position = split_string.index(search)\n",
    "    except ValueError:\n",
    "        position = -1  # 'AU' not found in the string\n",
    "    \n",
    "    return position\n",
    "\n",
    "def get_variant_reads(row, type, alt):\n",
    "\n",
    "    AU_index = get_position(row['FORMAT'],'AU')\n",
    "    CU_index = get_position(row['FORMAT'],'CU')\n",
    "    GU_index = get_position(row['FORMAT'],'GU')\n",
    "    TU_index = get_position(row['FORMAT'],'TU')\n",
    "                \n",
    "    if alt == \"A\":\n",
    "        variant_reads = row[type].split(\":\")[AU_index].split(',')\n",
    "        variant_reads = [int(x) for x in variant_reads] \n",
    "        total_variant_reads = sum(variant_reads)  # Sum the integer values\n",
    "    elif alt == \"C\":\n",
    "        variant_reads = row[type].split(\":\")[CU_index].split(',')\n",
    "        variant_reads = [int(x) for x in variant_reads] \n",
    "        total_variant_reads = sum(variant_reads)  # Sum the integer values\n",
    "    elif alt == \"G\":\n",
    "        variant_reads = row[type].split(\":\")[GU_index].split(',')\n",
    "        variant_reads = [int(x) for x in variant_reads] \n",
    "        total_variant_reads = sum(variant_reads)  # Sum the integer values\n",
    "    elif alt == \"T\":\n",
    "        variant_reads = row[type].split(\":\")[TU_index].split(',')\n",
    "        variant_reads = [int(x) for x in variant_reads] \n",
    "        total_variant_reads = sum(variant_reads)  # Sum the integer values\n",
    "    else:\n",
    "        raise Exception(\"Not recognized nucleotide\")\n",
    "\n",
    "    indices = [AU_index, CU_index, GU_index, TU_index]\n",
    "    total_reads = []\n",
    "    \n",
    "    for index in indices:\n",
    "        total_reads += row['TUMOR'].split(\":\")[index].split(',')\n",
    "                \n",
    "    total_reads = [int(x) for x in total_reads]\n",
    "    total_reads = sum(total_reads)\n",
    "    \n",
    "    return total_variant_reads, total_reads\n",
    "\n",
    "\n",
    "def convert_strelka_snv_to_tusv_ext_df(strelka_df, ascat_df, option):\n",
    "    \"\"\"\n",
    "    option can be phased or unphased\n",
    "    \"\"\"\n",
    "    tusv_ext_rows = []\n",
    "    strelka_df = strelka_df[strelka_df[\"FILTER\"] == \"PASS\"]\n",
    "    strelka_df.reset_index(inplace=True)\n",
    "    for index, row in strelka_df.iterrows():\n",
    "        chrom = row['#CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        if option == \"phased\":\n",
    "            raise Exception(\"Not implemented yet\")\n",
    "        elif option == \"unphased\":\n",
    "            data_types = [\"TUMOR\", \"NORMAL\"]\n",
    "            CNADJ_dict = {}\n",
    "            for type in data_types:  \n",
    "                total_variant_reads, total_reads = get_variant_reads(row, type, alt)\n",
    "                CN_tot = get_total_copy_number(ascat_df, chrom, pos)\n",
    "    \n",
    "                \n",
    "                CNADJ = (total_variant_reads)/(total_reads)*CN_tot\n",
    "                CNADJ = float(CNADJ)\n",
    "                CNADJ = format(CNADJ, '.2f')\n",
    "                CNADJ_dict[type] = CNADJ\n",
    "        else:\n",
    "            raise Exception(\"Unknown Option\")\n",
    "        qual = \".\"\n",
    "        filter_val = row[\"FILTER\"]\n",
    "        info = \".\"\n",
    "        variant_id = f\"snv{snv_count}\"\n",
    "        increment_snv()\n",
    "\n",
    "\n",
    "        format_field = 'GT:CNADJ'\n",
    "        tumor_data = \"1|1\" + \":\" + CNADJ_dict[\"TUMOR\"]\n",
    "        normal_data = \"0|0\" + \":\" + CNADJ_dict[\"NORMAL\"]\n",
    "        tusv_ext_rows.append([\n",
    "            chrom, pos, variant_id, ref, alt, qual, filter_val, info, format_field, tumor_data, normal_data\n",
    "        ])\n",
    "        \n",
    "    tusv_ext_df = pd.DataFrame(tusv_ext_rows, columns=[\n",
    "        '#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'TUMOR', 'NORMAL'\n",
    "    ])\n",
    "    return tusv_ext_df\n",
    "\n",
    "def read_manta_sv(input_file_path):\n",
    "    filtered_lines = []\n",
    "    with gzip.open(input_file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('##'):\n",
    "                filtered_lines.append(line)\n",
    "    filtered_content = ''.join(filtered_lines)\n",
    "    df_manta_sv = pd.read_csv(StringIO(filtered_content), sep='\\t')\n",
    "    df_manta_sv['#CHROM'] = df_manta_sv['#CHROM'].apply(convert_chromosome)\n",
    "    \n",
    "    return df_manta_sv\n",
    "\n",
    "\n",
    "\n",
    "def get_total_copy_number(df, chrom, pos):\n",
    "    \"\"\"\n",
    "    Gets the total copy number (major + minor) at a given chromosome and position.\n",
    "\n",
    "    Parameters:\n",
    "    df (pd.DataFrame): The DataFrame containing ASCAT data.\n",
    "    chrom (str or int): The chromosome number.\n",
    "    pos (int): The position on the chromosome.\n",
    "\n",
    "    Returns:\n",
    "    int or None: The total copy number at the given position, or None if the position is not covered.\n",
    "    \"\"\"\n",
    "    # Filter the DataFrame for the given chromosome\n",
    "    chr_df = df[df['chr'] == chrom]\n",
    "    \n",
    "    # Iterate over the rows to find the interval that contains the position\n",
    "    for _, row in chr_df.iterrows():\n",
    "        if row['startpos'] <= pos <= row['endpos']:\n",
    "            total_cn = row['nMajor'] + row['nMinor']\n",
    "            return total_cn\n",
    "    \n",
    "    # If no interval contains the position, return 2 assuming 2 normal alleles\n",
    "    return 2\n",
    "\n",
    "\n",
    "# Function to rename columns\n",
    "def rename_column(col_name):\n",
    "    last_split = col_name.split('-')[-1]\n",
    "    if last_split.startswith('T'):\n",
    "        return 'TUMOR'\n",
    "    elif last_split.startswith('N'):\n",
    "        return 'NORMAL'\n",
    "    return col_name\n",
    "# You may need to estimate CNAJ using CNVKit information tied to Manta output, and right hard conversion functions for DP and BDP\n",
    "\n",
    "\n",
    "def calculate_metrics(ascat_df, row):\n",
    "        # Ref then alt in the order list\n",
    "    data_columns = ['TUMOR', 'NORMAL']\n",
    "\n",
    "    for data in data_columns:\n",
    "        if row[\"FORMAT\"] == \"PR\":\n",
    "            PR = row[data]\n",
    "            \n",
    "            PR_ref = int(PR.split(\",\")[0])\n",
    "            PR_alt = int(PR.split(\",\")[1])\n",
    "            PR_tot = PR_ref + PR_alt\n",
    "            \n",
    "            SR_ref = 0\n",
    "            SR_alt = 0\n",
    "            SR_tot = 0\n",
    "        else:\n",
    "            try:\n",
    "        \n",
    "                PR = row[data].split(\":\")[0]\n",
    "                SR = row[data].split(\":\")[1]\n",
    "        \n",
    "                PR_ref = int(PR.split(\",\")[0])\n",
    "                PR_alt = int(PR.split(\",\")[1])\n",
    "                PR_tot = PR_ref + PR_alt\n",
    "        \n",
    "                \n",
    "                SR_ref = int(SR.split(\",\")[0])\n",
    "                SR_alt = int(SR.split(\",\")[1])\n",
    "                SR_tot = SR_ref + SR_alt\n",
    "    \n",
    "            except IndexError:\n",
    "                raise Exception(\"IndexError found\")\n",
    "\n",
    "        if data == 'TUMOR':\n",
    "            GT = \"1|1\"\n",
    "            CN_tot = get_total_copy_number(ascat_df, row['#CHROM'], row['POS'])\n",
    "            \n",
    "            CNADJ = (PR_alt + SR_alt)/(PR_tot + SR_tot)*CN_tot\n",
    "            CNADJ = float(CNADJ)\n",
    "            CNADJ_formatted = format(CNADJ, '.2f')\n",
    "\n",
    "            BDP = SR_alt\n",
    "            DP = SR_alt + PR_alt\n",
    "            variables = [GT, CNADJ_formatted, BDP, DP]\n",
    "            tumor_data = ':'.join(map(str, variables))\n",
    "        elif data == 'NORMAL':\n",
    "            GT = \"0|0\"\n",
    "            CN_tot = 2\n",
    "            CNADJ = (PR_ref + SR_ref)/(PR_tot + SR_tot)*CN_tot\n",
    "            CNADJ = float(CNADJ)\n",
    "            CNADJ_formatted = format(CNADJ, '.2f')\n",
    "            BDP = SR_ref\n",
    "            DP = SR_ref + PR_alt\n",
    "            variables = [GT, CNADJ_formatted, BDP, DP]\n",
    "            normal_data = ':'.join(map(str, variables))\n",
    "        else:\n",
    "            raise Exception(\"Unknown Datatype\")\n",
    "    return tumor_data, normal_data\n",
    "\n",
    "def rewrite_manta_results(df_manta_sv, ascat_df):\n",
    "\n",
    "    # Filter for SVs that pass all filters\n",
    "    df_manta_sv = df_manta_sv[df_manta_sv['FILTER'] == \"PASS\"]\n",
    "\n",
    "    \n",
    "    # Filter for BND types and paired breakpoints\n",
    "    bnd_df = df_manta_sv[df_manta_sv['INFO'].str.contains(\"SVTYPE=BND\")]\n",
    "\n",
    "    # Split the INFO column into a dictionary for easy access\n",
    "    bnd_df['INFO_DICT'] = bnd_df['INFO'].apply(lambda x: dict(item.split('=') for item in x.split(';') if '=' in item))\n",
    "\n",
    "    # Filter out unpaired BNDs\n",
    "    mate_ids = bnd_df['INFO_DICT'].apply(lambda x: x.get('MATEID', None))\n",
    "    paired_bnd_df = bnd_df[bnd_df['ID'].isin(mate_ids.values)]\n",
    "\n",
    "\n",
    "    # Prepare new VCF columns\n",
    "    vcf_columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'TUMOR', 'NORMAL']\n",
    "    vcf_df = pd.DataFrame(columns=vcf_columns)\n",
    "    \n",
    "    # Create a dictionary to store the paired rows\n",
    "    paired_rows = {}\n",
    "\n",
    "    # Renaming column heads for ease of access downstream to TUMOR and NORMAL\n",
    "    paired_bnd_df.rename(columns=lambda x: rename_column(x), inplace=True)\n",
    "    print(paired_bnd_df['FORMAT'].unique())\n",
    "    for index, row in paired_bnd_df.iterrows():\n",
    "        mate_id = row['INFO_DICT']['MATEID']\n",
    "        paired_rows[row['ID']] = row\n",
    "        if mate_id in paired_rows:\n",
    "\n",
    "\n",
    "            \n",
    "            mate_row = paired_rows[mate_id]\n",
    "\n",
    "\n",
    "            # Create the ALT field\n",
    "            alt = row[\"ALT\"]\n",
    "            alt = str(alt).replace('chr','')\n",
    "            alt = alt.replace(row['REF'],'')\n",
    "            alt = alt.replace('X','23')\n",
    "            #Recalculate ID here for row and mate_row\n",
    "            row_sv_ID = f\"sv{sv_count}\"\n",
    "            increment_sv()\n",
    "\n",
    "            mate_row_sv_ID = f\"sv{sv_count}\"\n",
    "            increment_sv()\n",
    "            \n",
    "            # Create the INFO field\n",
    "            info = f\"MATEID={mate_row_sv_ID};SVTYPE=BND\"\n",
    "\n",
    "            # Create the FORMAT fields \n",
    "            format_field = \"GT:CNADJ:BDP:DP\"\n",
    "\n",
    "\n",
    "            tumor_data, normal_data = calculate_metrics(ascat_df, row)\n",
    "\n",
    "\n",
    "            # Append the rows to the new DataFrame\n",
    "            vcf_df = vcf_df.append({\n",
    "                '#CHROM': row['#CHROM'], 'POS': row['POS'], 'ID': row_sv_ID, 'REF': '.',\n",
    "                'ALT': alt, 'QUAL': row['QUAL'], 'FILTER': row['FILTER'], 'INFO': info,\n",
    "                'FORMAT': format_field, 'TUMOR': tumor_data, 'NORMAL': normal_data\n",
    "            }, ignore_index=True)\n",
    "\n",
    "\n",
    "            # Create the ALT field\n",
    "            alt_mate = mate_row[\"ALT\"]\n",
    "            alt_mate = str(alt_mate).replace('chr','')\n",
    "            alt_mate = alt_mate.replace(mate_row['REF'],'')\n",
    "            alt_mate = alt_mate.replace('X','23')\n",
    "\n",
    "\n",
    "            info_mate = f\"MATEID={row_sv_ID};SVTYPE=BND\"\n",
    "\n",
    "\n",
    "            tumor_data, normal_data = calculate_metrics(ascat_df, mate_row)\n",
    "\n",
    "            \n",
    "            vcf_df = vcf_df.append({\n",
    "                '#CHROM': mate_row['#CHROM'], 'POS': mate_row['POS'], 'ID': mate_row_sv_ID, 'REF': '.',\n",
    "                'ALT': alt_mate, 'QUAL': mate_row['QUAL'], 'FILTER': mate_row['FILTER'], 'INFO': info_mate,\n",
    "                'FORMAT': format_field, 'TUMOR': tumor_data, 'NORMAL': normal_data\n",
    "            }, ignore_index=True)\n",
    "\n",
    "    return vcf_df\n",
    "    \n",
    "def main():\n",
    "    global cnv_count, sv_count, snv_count\n",
    "    cnv_count = 1\n",
    "    sv_count = 1\n",
    "    snv_count = 1\n",
    "    variant_calling_folder = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/AER5_wes_variant_calling/'\n",
    "    output_folder = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/compiled_input/'\n",
    "    sample_name = 'AUR-AER5-TTM4_vs_AUR-AER5-NT2'\n",
    "    frozen_samples = ['AUR-AER5-TTM4_vs_AUR-AER5-NT2', 'AUR-AER5-TTM5_vs_AUR-AER5-NT2', 'AUR-AER5-TTP2_vs_AUR-AER5-NT2']\n",
    "    # ASCAT_JOINT_MUTECT\n",
    "\n",
    "    # Path to the GRCh38 reference genome file\n",
    "    reference_genome_path = \"/bgfs/alee/LO_LAB/General/References/GCF_GRCh38/GCF_000001405.40_GRCh38.p14_genomic.fna\"\n",
    "\n",
    "    \n",
    "    for sample_name in frozen_samples:\n",
    "        print(sample_name)\n",
    "        strelka_snv_path = os.path.join(variant_calling_folder, 'strelka', sample_name, sample_name + '.strelka.somatic_snvs.vcf.gz')\n",
    "        strelka_indel_path = os.path.join(variant_calling_folder, 'strelka', sample_name, sample_name + '.strelka.somatic_indels.vcf.gz')\n",
    "        manta_sv_path = os.path.join(variant_calling_folder, 'manta', sample_name, sample_name + '.manta.somatic_sv.vcf.gz')\n",
    "        ascat_cnv_path = os.path.join(variant_calling_folder, 'ascat', sample_name, sample_name + '.cnvs.txt')\n",
    "    \n",
    "    \n",
    "        output_file_path = init_vcf(output_folder, sample_name)\n",
    "    \n",
    "        \n",
    "        ascat_df = load_ascat_cnv(ascat_cnv_path)\n",
    "        vcf_df = convert_ascat_to_vcf_df(ascat_df, reference_genome_path)\n",
    "        write_df_to_vcf(vcf_df, output_file_path)\n",
    "    \n",
    "    \n",
    "        df_manta_sv = read_manta_sv(manta_sv_path)\n",
    "        df_processed_manta = rewrite_manta_results(df_manta_sv, ascat_df)\n",
    "        write_df_to_vcf(df_processed_manta, output_file_path)\n",
    "    \n",
    "        df_snv = read_strelka(strelka_snv_path)\n",
    "        tusv_ext_snv = convert_strelka_snv_to_tusv_ext_df(df_snv, ascat_df, \"unphased\")\n",
    "        write_df_to_vcf(tusv_ext_snv, output_file_path)\n",
    "\n",
    "    # TODO: write assertion checks for the final document. \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8861ffe0-598a-4a04-9ce4-b389343f9625",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d651b20-a8d3-4715-a74d-0672ae4e5922",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13bb60d0-0003-407e-b7d9-01e544c05eed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2b19af-48b6-44df-93ca-17ff9170873c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4e5668-b0f5-4ddf-aa25-2db8586684ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "87d385f6-f859-421f-86d7-7827e31bf226",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BreakEnd('1', 185147263, '+', '-', '', True)\n",
      "BreakEnd('1', 185147263, '+', '-', '', True)\n",
      "BreakEnd('3', 116437976, '+', '+', '', True)\n",
      "BreakEnd('3', 116437976, '+', '+', '', True)\n",
      "BreakEnd('2', 43710295, '+', '+', 'T', True)\n",
      "BreakEnd('2', 43710295, '+', '+', 'T', True)\n",
      "BreakEnd('6', 41633110, '+', '+', 'A', True)\n",
      "BreakEnd('6', 41633110, '+', '+', 'A', True)\n",
      "BreakEnd('4', 156027395, '+', '+', '', True)\n",
      "BreakEnd('4', 156027395, '+', '+', '', True)\n",
      "BreakEnd('7', 4920148, '+', '+', '', True)\n",
      "BreakEnd('7', 4920148, '+', '+', '', True)\n",
      "BreakEnd('6', 72182816, '+', '-', '', True)\n",
      "BreakEnd('6', 72182816, '+', '-', '', True)\n",
      "BreakEnd('7', 65845265, '+', '-', '', True)\n",
      "BreakEnd('7', 65845265, '+', '-', '', True)\n",
      "BreakEnd('4', 1701063, '-', '+', 'AACA', True)\n",
      "BreakEnd('4', 1701063, '-', '+', 'AACA', True)\n",
      "BreakEnd('7', 100205251, '+', '-', 'GGAGACA', True)\n",
      "BreakEnd('7', 100205251, '+', '-', 'GGAGACA', True)\n",
      "BreakEnd('1', 145831220, '+', '-', '', True)\n",
      "BreakEnd('1', 145831220, '+', '-', '', True)\n",
      "BreakEnd('7', 128030113, '+', '+', '', True)\n",
      "BreakEnd('7', 128030113, '+', '+', '', True)\n",
      "BreakEnd('7', 141919730, '+', '+', '', True)\n",
      "BreakEnd('7', 141919730, '+', '+', '', True)\n",
      "BreakEnd('7', 141919817, '+', '+', '', True)\n",
      "BreakEnd('7', 141919817, '+', '+', '', True)\n",
      "BreakEnd('6', 1710928, '+', '-', 'A', True)\n",
      "BreakEnd('6', 1710928, '+', '-', 'A', True)\n",
      "BreakEnd('8', 101691013, '-', '+', 'A', True)\n",
      "BreakEnd('8', 101691013, '-', '+', 'A', True)\n",
      "BreakEnd('6', 108179332, '+', '+', '', True)\n",
      "BreakEnd('6', 108179332, '+', '+', '', True)\n",
      "BreakEnd('10', 103571868, '+', '-', 'A', True)\n",
      "BreakEnd('10', 103571868, '+', '-', 'A', True)\n",
      "BreakEnd('11', 111543864, '+', '+', '', True)\n",
      "BreakEnd('11', 111543864, '+', '+', '', True)\n",
      "BreakEnd('11', 111543950, '+', '+', '', True)\n",
      "BreakEnd('11', 111543950, '+', '+', '', True)\n",
      "BreakEnd('2', 179100236, '+', '-', '', True)\n",
      "BreakEnd('2', 179100236, '+', '-', '', True)\n",
      "BreakEnd('12', 2797743, '+', '-', '', True)\n",
      "BreakEnd('12', 2797743, '+', '-', '', True)\n",
      "BreakEnd('2', 39309423, '+', '+', '', True)\n",
      "BreakEnd('2', 39309423, '+', '+', '', True)\n",
      "BreakEnd('12', 3435694, '+', '-', '', True)\n",
      "BreakEnd('12', 3435694, '+', '-', '', True)\n",
      "BreakEnd('8', 129556957, '+', '-', 'T', True)\n",
      "BreakEnd('8', 129556957, '+', '-', 'T', True)\n",
      "BreakEnd('12', 109052558, '-', '+', 'T', True)\n",
      "BreakEnd('12', 109052558, '-', '+', 'T', True)\n",
      "BreakEnd('11', 69538479, '-', '-', 'T', True)\n",
      "BreakEnd('11', 69538479, '-', '-', 'T', True)\n",
      "BreakEnd('12', 116276619, '+', '-', '', True)\n",
      "BreakEnd('12', 116276619, '+', '-', '', True)\n",
      "BreakEnd('15', 38458804, '+', '-', '', True)\n",
      "BreakEnd('15', 38458804, '+', '-', '', True)\n",
      "BreakEnd('15', 38458904, '+', '-', '', True)\n",
      "BreakEnd('15', 38458904, '+', '-', '', True)\n",
      "BreakEnd('15', 42404857, '+', '-', '', True)\n",
      "BreakEnd('15', 42404857, '+', '-', '', True)\n",
      "BreakEnd('15', 42404927, '+', '-', '', True)\n",
      "BreakEnd('15', 42404927, '+', '-', '', True)\n",
      "BreakEnd('4', 42494266, '+', '-', '', True)\n",
      "BreakEnd('4', 42494266, '+', '-', '', True)\n",
      "BreakEnd('15', 64681462, '+', '+', '', True)\n",
      "BreakEnd('15', 64681462, '+', '+', '', True)\n",
      "BreakEnd('5', 177260026, '+', '+', 'A', True)\n",
      "BreakEnd('5', 177260026, '+', '+', 'A', True)\n",
      "BreakEnd('15', 70034648, '+', '+', 'T', True)\n",
      "BreakEnd('15', 70034648, '+', '+', 'T', True)\n",
      "BreakEnd('13', 21219676, '+', '-', '', True)\n",
      "BreakEnd('13', 21219676, '+', '-', '', True)\n",
      "BreakEnd('15', 90223620, '+', '-', '', True)\n",
      "BreakEnd('15', 90223620, '+', '-', '', True)\n",
      "BreakEnd('5', 179615621, '+', '-', '', True)\n",
      "BreakEnd('5', 179615621, '+', '-', '', True)\n",
      "BreakEnd('16', 9016911, '+', '-', '', True)\n",
      "BreakEnd('16', 9016911, '+', '-', '', True)\n",
      "BreakEnd('1', 192594874, '+', '-', '', True)\n",
      "BreakEnd('1', 192594874, '+', '-', '', True)\n",
      "BreakEnd('18', 23536706, '+', '+', '', True)\n",
      "BreakEnd('18', 23536706, '+', '+', '', True)\n",
      "BreakEnd('5', 91350618, '+', '+', '', True)\n",
      "BreakEnd('5', 91350618, '+', '+', '', True)\n",
      "BreakEnd('18', 23536869, '+', '-', '', True)\n",
      "BreakEnd('18', 23536869, '+', '-', '', True)\n",
      "BreakEnd('1', 204276755, '+', '-', '', True)\n",
      "BreakEnd('1', 204276755, '+', '-', '', True)\n",
      "BreakEnd('18', 23749366, '+', '+', '', True)\n",
      "BreakEnd('18', 23749366, '+', '+', '', True)\n",
      "BreakEnd('17', 1058053, '+', '-', '', True)\n",
      "BreakEnd('17', 1058053, '+', '-', '', True)\n",
      "BreakEnd('18', 29803220, '+', '-', '', True)\n",
      "BreakEnd('18', 29803220, '+', '-', '', True)\n",
      "BreakEnd('18', 31524777, '+', '-', '', True)\n",
      "BreakEnd('18', 31524777, '+', '-', '', True)\n",
      "BreakEnd('18', 31524837, '+', '-', '', True)\n",
      "BreakEnd('18', 31524837, '+', '-', '', True)\n",
      "BreakEnd('2', 231000448, '+', '-', '', True)\n",
      "BreakEnd('2', 231000448, '+', '-', '', True)\n",
      "BreakEnd('19', 16480348, '+', '-', '', True)\n",
      "BreakEnd('19', 16480348, '+', '-', '', True)\n",
      "BreakEnd('7', 105145372, '+', '-', '', True)\n",
      "BreakEnd('7', 105145372, '+', '-', '', True)\n",
      "BreakEnd('19', 24008380, '+', '-', '', True)\n",
      "BreakEnd('19', 24008380, '+', '-', '', True)\n",
      "BreakEnd('12', 109444319, '+', '-', '', True)\n",
      "BreakEnd('12', 109444319, '+', '-', '', True)\n",
      "BreakEnd('19', 30440213, '-', '-', 'T', True)\n",
      "BreakEnd('19', 30440213, '-', '-', 'T', True)\n",
      "BreakEnd('6', 73451069, '+', '-', '', True)\n",
      "BreakEnd('6', 73451069, '+', '-', '', True)\n",
      "BreakEnd('19', 47355147, '+', '-', '', True)\n",
      "BreakEnd('19', 47355147, '+', '-', '', True)\n",
      "BreakEnd('13', 76281842, '+', '+', '', True)\n",
      "BreakEnd('13', 76281842, '+', '+', '', True)\n",
      "BreakEnd('20', 19607148, '+', '-', 'A', True)\n",
      "BreakEnd('20', 19607148, '+', '-', 'A', True)\n",
      "BreakEnd('11', 66691642, '+', '+', '', True)\n",
      "BreakEnd('11', 66691642, '+', '+', '', True)\n",
      "BreakEnd('20', 38262626, '+', '+', '', True)\n",
      "BreakEnd('20', 38262626, '+', '+', '', True)\n",
      "BreakEnd('7', 3654952, '+', '+', '', True)\n",
      "BreakEnd('7', 3654952, '+', '+', '', True)\n",
      "BreakEnd('22', 19142575, '+', '-', '', True)\n",
      "BreakEnd('22', 19142575, '+', '-', '', True)\n",
      "BreakEnd('12', 120573902, '+', '-', 'T', True)\n",
      "BreakEnd('12', 120573902, '+', '-', 'T', True)\n",
      "BreakEnd('22', 19801514, '-', '+', 'T', True)\n",
      "BreakEnd('22', 19801514, '-', '+', 'T', True)\n",
      "BreakEnd('X', 41229284, '+', '-', '', True)\n",
      "BreakEnd('X', 41229284, '+', '-', '', True)\n",
      "BreakEnd('X', 69402166, '+', '-', '', True)\n",
      "BreakEnd('X', 69402166, '+', '-', '', True)\n",
      "BreakEnd('X', 41229285, '+', '+', '', True)\n",
      "BreakEnd('X', 41229285, '+', '+', '', True)\n",
      "BreakEnd('X', 69402180, '+', '+', '', True)\n",
      "BreakEnd('X', 69402180, '+', '+', '', True)\n",
      "BreakEnd('X', 71148822, '+', '-', '', True)\n",
      "BreakEnd('X', 71148822, '+', '-', '', True)\n",
      "BreakEnd('X', 71148891, '+', '-', '', True)\n",
      "BreakEnd('X', 71148891, '+', '-', '', True)\n",
      "BreakEnd('X', 119603538, '+', '-', '', True)\n",
      "BreakEnd('X', 119603538, '+', '-', '', True)\n",
      "BreakEnd('X', 119604938, '+', '-', '', True)\n",
      "BreakEnd('X', 119604938, '+', '-', '', True)\n"
     ]
    }
   ],
   "source": [
    "import vcfpy\n",
    "\n",
    "\n",
    "def is_cnv_record(rec):\n",
    "    return rec.ID[0][0:3] == 'cnv'\n",
    "\n",
    "\n",
    "def is_sv_record(rec):\n",
    "    return rec.ID[0][0:2] == 'sv'\n",
    "\n",
    "\n",
    "def is_snv_record(rec):\n",
    "    return rec.ID[0][0:3] == 'snv'\n",
    "\n",
    "\n",
    "vcf_reader =  vcfpy.Reader.from_path('/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/compiled_input/AUR-AER5-TTM4_vs_AUR-AER5-NT2.vcf')\n",
    "\n",
    "for rec in vcf_reader:\n",
    "    if is_sv_record(rec):\n",
    "        print(rec.ALT[0])\n",
    "        print(rec.ALT[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "c9af8994-e7f2-4841-bf66-4c89b1ef536a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are more SVs in one or more samples than your SV Upperbound, increasing SV_UB to 184, and C to 224\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "output_folder = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/compiled_input/'\n",
    "\n",
    "\n",
    "sampleList = os.listdir(output_folder)\n",
    "sv_ub = 80\n",
    "max_sv = -1\n",
    "for i, sample in enumerate(sampleList):\n",
    "    sv_count = 0\n",
    "    input_vcf_file = output_folder  + sample\n",
    "    reader =  vcfpy.Reader.from_path(input_vcf_file)\n",
    "    for rec in reader:\n",
    "        if is_sv_record(rec):\n",
    "            sv_count += 1\n",
    "    max_sv = max(max_sv, sv_count)\n",
    "if max_sv >= sv_ub:\n",
    "    print(f\"There are more SVs in one or more samples than your SV Upperbound, increasing SV_UB to {max_sv}, and C to {max_sv+40}\")\n",
    "    sv_ub = max_sv\n",
    "    const = max_sv+40\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "928b6e8b-4cfb-4403-b1cc-7b5ea81ff76f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 01: 248956422 bp\n",
      "Chromosome 02: 242193529 bp\n",
      "Chromosome 03: 198295559 bp\n",
      "Chromosome 04: 190214555 bp\n",
      "Chromosome 05: 181538259 bp\n",
      "Chromosome 06: 170805979 bp\n",
      "Chromosome 07: 159345973 bp\n",
      "Chromosome 08: 145138636 bp\n",
      "Chromosome 09: 138394717 bp\n",
      "Chromosome 10: 133797422 bp\n",
      "Chromosome 11: 135086622 bp\n",
      "Chromosome 12: 133275309 bp\n",
      "Chromosome 13: 114364328 bp\n",
      "Chromosome 14: 107043718 bp\n",
      "Chromosome 15: 101991189 bp\n",
      "Chromosome 16: 90338345 bp\n",
      "Chromosome 17: 83257441 bp\n",
      "Chromosome 18: 80373285 bp\n",
      "Chromosome 19: 58617616 bp\n",
      "Chromosome 20: 64444167 bp\n",
      "Chromosome 21: 46709983 bp\n",
      "Chromosome 22: 50818468 bp\n",
      "Chromosome 23: 156040895 bp\n",
      "Chromosome MT: 16569 bp\n",
      "Chromosome 24: 57227415 bp\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Path to the GRCh38 reference genome file\n",
    "reference_genome_path = \"/bgfs/alee/LO_LAB/General/References/GCF_GRCh38/GCF_000001405.40_GRCh38.p14_genomic.fna\"\n",
    "\n",
    "# Dictionary to store chromosome lengths\n",
    "chromosome_lengths = {}\n",
    "\n",
    "# Parse the reference genome file\n",
    "for record in SeqIO.parse(reference_genome_path, \"fasta\"):\n",
    "    chromosome = record.id\n",
    "    length = len(record.seq)\n",
    "    if chromosome.startswith(\"NC_0000\") or chromosome.startswith(\"NC_012920.1\"):\n",
    "        chromosome_number = chromosome.split('.')[0].replace(\"NC_0000\", \"\").replace(\"NC_012920\", \"MT\")\n",
    "        chromosome_lengths[chromosome_number] = length\n",
    "\n",
    "# Sort chromosome lengths based on chromosome number\n",
    "sorted_chromosome_lengths = dict(sorted(chromosome_lengths.items(), key=lambda item: int(item[0].replace(\"MT\", \"25\").replace(\"X\", \"23\").replace(\"Y\", \"24\"))))\n",
    "\n",
    "# Print the lengths of the main chromosomes\n",
    "for chromosome, length in sorted_chromosome_lengths.items():\n",
    "    print(f\"Chromosome {chromosome}: {length} bp\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66902b23-a47b-440a-a0e5-bd488c04d464",
   "metadata": {},
   "source": [
    "import vcf\n",
    "\n",
    "vcf_reader = vcf.Reader(open('/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/compiled_input/AUR-AER5-TTM4_vs_AUR-AER5-NT2.vcf', 'r'))\n",
    "\n",
    "for record in vcf_reader:\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbeb034b-ff4a-4177-8941-d7d0badddb76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to add an option so SNVs are given SNV IDs but strelka indels are given SV ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "174cd4e2-fb66-485b-b9b3-fcafa3ced517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #CHROM    POS ID REF ALT QUAL  FILTER  \\\n",
      "0   chr1  10120  .   T   C    .  LowEVS   \n",
      "1   chr1  10257  .   A   C    .  LowEVS   \n",
      "2   chr1  10330  .   C   A    .  LowEVS   \n",
      "3   chr1  14063  .   G   A    .  LowEVS   \n",
      "4   chr1  14653  .   C   T    .  LowEVS   \n",
      "\n",
      "                                                INFO  \\\n",
      "0  DP=1080;MQ=12.08;MQ0=874;NT=ref;QSS=8;QSS_NT=8...   \n",
      "1  DP=190;MQ=16.38;MQ0=92;NT=ref;QSS=3;QSS_NT=3;R...   \n",
      "2  DP=212;MQ=21.15;MQ0=89;NT=ref;QSS=1;QSS_NT=1;R...   \n",
      "3  DP=37;MQ=18.01;MQ0=25;NT=ref;QSS=1;QSS_NT=1;Re...   \n",
      "4  DP=45;MQ=29.11;MQ0=14;NT=ref;QSS=4;QSS_NT=4;Re...   \n",
      "\n",
      "                         FORMAT                         NORMAL  \\\n",
      "0  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  48:18:1:0:2,19:0,2:0,2:28,112   \n",
      "1  DP:FDP:SDP:SUBDP:AU:CU:GU:TU      12:7:1:0:5,41:0,5:0,0:0,0   \n",
      "2  DP:FDP:SDP:SUBDP:AU:CU:GU:TU     26:19:6:0:0,5:7,23:0,1:0,1   \n",
      "3  DP:FDP:SDP:SUBDP:AU:CU:GU:TU       3:0:0:0:0,0:0,0:3,14:0,0   \n",
      "4  DP:FDP:SDP:SUBDP:AU:CU:GU:TU     14:0:0:0:0,0:12,18:0,0:2,3   \n",
      "\n",
      "                          TUMOR  chr                          data  \n",
      "0  48:17:0:0:2,14:3,4:0,5:26,91    1  48:17:0:0:2,14:3,4:0,5:26,91  \n",
      "1   19:14:2:0:3,50:2,10:0,0:0,0    1   19:14:2:0:3,50:2,10:0,0:0,0  \n",
      "2   29:22:1:0:2,11:5,24:0,0:0,0    1   29:22:1:0:2,11:5,24:0,0:0,0  \n",
      "3      4:0:0:0:2,2:0,0:2,21:0,0    1      4:0:0:0:2,2:0,0:2,21:0,0  \n",
      "4    15:0:0:0:0,0:13,22:0,0:2,2    1    15:0:0:0:0,0:13,22:0,0:2,2  \n",
      "(1305762, 13)\n",
      "(27859, 13)\n",
      "  #CHROM     POS      ID REF ALT QUAL FILTER INFO    FORMAT  TUMOR NORMAL\n",
      "0   chr1  202192  sv0000   T   C    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "1   chr1  791204  sv0001   T   C    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "2   chr1  795801  sv0002   G   A    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "3   chr1  818228  sv0003   T   A    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n",
      "4   chr1  881300  sv0004   G   T    .   PASS    .  GT:CNADJ  0|1:0  0|0:0\n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in Strelka indel files to test\n",
    "\n",
    "test_strelka_indel_path = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/strelka/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.strelka.somatic_snvs.vcf.gz'\n",
    "\n",
    "df_indel = read_strelka(test_strelka_indel_path)\n",
    "\n",
    "\n",
    "print(df_indel.head())\n",
    "tusv_ext_indel = convert_strelka_to_tusv_ext_df(df_indel, \"INDEL\")\n",
    "print(tusv_ext_indel.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4e3464e2-58c7-47ed-bd5c-2f7d79896fec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14497, 13)\n",
      "(5411, 13)\n",
      "  #CHROM      POS       ID REF               ALT QUAL FILTER INFO    FORMAT  \\\n",
      "0   chr1  1056668  snv0000   C               CGT    .   PASS    .  GT:CNADJ   \n",
      "1   chr1  1423775  snv0001   C                CT    .   PASS    .  GT:CNADJ   \n",
      "2   chr1  1509914  snv0002  TA                 T    .   PASS    .  GT:CNADJ   \n",
      "3   chr1  2900016  snv0003   C                CG    .   PASS    .  GT:CNADJ   \n",
      "4   chr1  3329041  snv0004   G  GCCTTCTCCCTGGGCC    .   PASS    .  GT:CNADJ   \n",
      "\n",
      "   TUMOR NORMAL  \n",
      "0  0|1:0  0|0:0  \n",
      "1  0|1:0  0|0:0  \n",
      "2  0|1:0  0|0:0  \n",
      "3  0|1:0  0|0:0  \n",
      "4  0|1:0  0|0:0  \n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in Strelka snv files to test.\n",
    "import sys\n",
    "\n",
    "test_strelka_snv_path = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/strelka/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.strelka.somatic_indels.vcf.gz'\n",
    "\n",
    "df_snv = read_strelka(test_strelka_snv_path)\n",
    "\n",
    "tusv_ext_snv = convert_strelka_to_tusv_ext_df(df_snv, \"SNV\")\n",
    "print(tusv_ext_snv.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f9a2a282-781c-4c9e-9d2b-c22101d9440e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chromosome    start       end gene      log2     depth  probes     weight  \\\n",
      "0       chr1    14034    351968    -  0.098452  14.55930      58    52.1960   \n",
      "1       chr1   351968    459979    - -0.413539   7.87716      27    19.1118   \n",
      "2       chr1   459979   1264807    - -0.022918  13.51170     180   170.4730   \n",
      "3       chr1  1264807   1333091    -  0.322362  17.72340      17    16.8359   \n",
      "4       chr1  1333091  16558479    - -0.011359  13.51330    3752  3682.8400   \n",
      "\n",
      "      ci_lo     ci_hi  \n",
      "0  0.080594  0.119033  \n",
      "1 -0.444376 -0.386697  \n",
      "2 -0.031888 -0.010873  \n",
      "3  0.294485  0.344509  \n",
      "4 -0.013140 -0.009817  \n",
      "  #CHROM      POS     ID REF    ALT QUAL FILTER                    INFO  \\\n",
      "0   chr1    14034  cnv00   .  <CNV>    .   PASS    END=351968;IMPRECISE   \n",
      "1   chr1   351968  cnv01   .  <CNV>    .   PASS    END=459979;IMPRECISE   \n",
      "2   chr1   459979  cnv02   .  <CNV>    .   PASS   END=1264807;IMPRECISE   \n",
      "3   chr1  1264807  cnv03   .  <CNV>    .   PASS   END=1333091;IMPRECISE   \n",
      "4   chr1  1333091  cnv04   .  <CNV>    .   PASS  END=16558479;IMPRECISE   \n",
      "\n",
      "  FORMAT                   TUMOR   NORMAL  \n",
      "0  GT:CN   1|1:0.0984518,14.5593  0|0:1,1  \n",
      "1  GT:CN   1|1:-0.413539,7.87716  0|0:1,1  \n",
      "2  GT:CN  1|1:-0.0229176,13.5117  0|0:1,1  \n",
      "3  GT:CN    1|1:0.322362,17.7234  0|0:1,1  \n",
      "4  GT:CN  1|1:-0.0113593,13.5133  0|0:1,1  \n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in CNVKit .cns files to test.\n",
    "\n",
    "test_cnvkit_cns_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/cnvkit/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4.cns'\n",
    "\n",
    "cnvkit_df = load_cnvkit_cns(test_cnvkit_cns_file)\n",
    "print(cnvkit_df.head())\n",
    "\n",
    "vcf_df = cnvkit_to_df(cnvkit_df)\n",
    "print(vcf_df.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e89c704e-7841-4e28-bfba-2e5466d8d68a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #CHROM       POS                               ID REF                 ALT  \\\n",
      "0   chr1   3722167       MantaBND:11184:1:3:0:0:0:1   A     A]chr1:3722443]   \n",
      "1   chr1   3722434       MantaBND:11184:1:3:0:0:0:0   A     A]chr1:3722176]   \n",
      "2   chr1  10413332       MantaBND:73506:1:1:6:0:0:1   T    T]chr1:10413436]   \n",
      "3   chr1  10413428       MantaBND:73506:1:1:6:0:0:0   T    T]chr1:10413340]   \n",
      "4   chr1  15349115       MantaBND:87210:0:0:1:0:0:1   A    A]chr1:15349202]   \n",
      "5   chr1  15349194       MantaBND:87210:0:0:1:0:0:0   G    G]chr1:15349123]   \n",
      "6   chr1  16996026   MantaBND:1:97014:97014:1:0:0:0   G    [chr1:16996118[G   \n",
      "7   chr1  16996113   MantaBND:1:97014:97014:1:0:0:1   C    [chr1:16996031[C   \n",
      "8   chr1  20792077   MantaBND:1:31559:31577:0:0:0:1   G  GT]chr18:37273578]   \n",
      "9   chr1  20867797  MantaDUP:TANDEM:12529:1:7:0:1:0   A        <DUP:TANDEM>   \n",
      "\n",
      "  QUAL           FILTER                                               INFO  \\\n",
      "0    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:11184:1:3:0:0:0:0;C...   \n",
      "1    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:11184:1:3:0:0:0:1;C...   \n",
      "2    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:73506:1:1:6:0:0:0;C...   \n",
      "3    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:73506:1:1:6:0:0:1;C...   \n",
      "4    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:87210:0:0:1:0:0:0;C...   \n",
      "5    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:87210:0:0:1:0:0:1;C...   \n",
      "6    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:1:97014:97014:1:0:0...   \n",
      "7    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:1:97014:97014:1:0:0...   \n",
      "8    .  MinSomaticScore  SVTYPE=BND;MATEID=MantaBND:1:31559:31577:0:0:0...   \n",
      "9    .  MinSomaticScore  END=117102440;SVTYPE=DUP;SVLEN=96234643;CIPOS=...   \n",
      "\n",
      "  FORMAT AUR-AER5_AUR-AER5-NT2 AUR-AER5_AUR-AER5-TTM4  \n",
      "0  PR:SR            99,0:207,0            121,2:288,3  \n",
      "1  PR:SR            99,0:207,0            121,2:288,3  \n",
      "2  PR:SR             16,0:57,0              20,2:68,2  \n",
      "3  PR:SR             16,0:57,0              20,2:68,2  \n",
      "4  PR:SR            60,0:146,0            100,2:240,2  \n",
      "5  PR:SR            60,0:146,0            100,2:240,2  \n",
      "6  PR:SR           166,0:437,0            199,2:504,3  \n",
      "7  PR:SR           166,0:437,0            199,2:504,3  \n",
      "8  PR:SR           103,0:252,0            183,2:418,3  \n",
      "9  PR:SR              1,0:16,0               0,0:15,3  \n"
     ]
    }
   ],
   "source": [
    "# Test box for reading in Manta .vcf files to determine format.\n",
    "\n",
    "test_manta_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/wholebody_phylo/AER5_results/variant_calling/manta/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.manta.somatic_sv.vcf.gz'\n",
    "test_organoid_wes_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/nfcore/achang4_bridgesPSC/sarek/Part2_FASTQ2SAREK/results/variant_calling/manta/123_organoid_vs_123_blood/123_organoid_vs_123_blood.manta.somatic_sv.vcf.gz'\n",
    "test_manta_wes_file = '/bgfs/alee/LO_LAB/Personal/Alexander_Chang/alc376/TUSV-ext/variant_calling/manta/AUR-AER5-TTM4_vs_AUR-AER5-NT2/AUR-AER5-TTM4_vs_AUR-AER5-NT2.manta.somatic_sv.vcf.gz'\n",
    "\n",
    "df_processed_manta = read_manta_sv(test_manta_wes_file)\n",
    "\n",
    "print(df_processed_manta.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c1427d-4c89-4117-8ca4-b25b6d4a540e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proceed with WES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42f9386-5616-4fcd-b384-f3a39a03a1f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TUSV-ext",
   "language": "python",
   "name": "tusv_ext"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
